runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/FileWatcherTask.java

```diff
@@ -0,0 +1,147 @@
+package io.aklivity.zilla.runtime.engine.internal.registry;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static org.agrona.LangUtil.rethrowUnchecked;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+
+public class FileWatcherTask extends WatcherTask
+{
+    private final Map<Path, byte[]> configHashes;
+    private WatchService watchService;
+
+    public FileWatcherTask(BiConsumer<URL, String> configChangeListener)
+    {
+        super(configChangeListener);
+        this.configHashes = new HashMap<>();
+        try
+        {
+            watchService = FileSystems.getDefault().newWatchService();
+        }
+        catch (IOException ex)
+        {
+            rethrowUnchecked(ex);
+        }
+
+    }
+
+    @Override
+    public boolean run()
+    {
+        try
+        {
+            while (true)
+            {
+                try
+                {
+                    final WatchKey key = watchService.take();
+
+                    for (WatchEvent<?> event : key.pollEvents())
+                    {
+                        final Path changed = ((Path) key.watchable()).resolve((Path) event.context());
+                        if (configHashes.containsKey(changed))
+                        {
+                            String configText = readConfigText(configURLs.get(changed));
+                            byte[] newConfigHash = computeHash(configText);
+                            if (!Arrays.equals(configHashes.get(changed), newConfigHash))
+                            {
+                                configHashes.put(changed, newConfigHash);
+                                configChangeListener.accept(configURLs.get(changed), configText);
+                            }
+                        }
+                    }
+                    key.reset();
+                }
+                catch (InterruptedException ex)
+                {
+                    watchService.close();
```

Comment:
Instead of using an explicit `Thread.interrupt()` to escape the `take()` method and react by closing the watch service, can we have a `close()` method on the watcher task that calls `watcherService.close()` and triggers `ClosedWatchServiceException` to escape the `take()` method?

Reply:
We already discussed this: https://github.com/aklivity/zilla/pull/141/commits/d9d197ad2ffb2dbb87770810be015d5a49e5f2f1#r1082477238
Short answer: it's not enough

Reply:
Yeah, I recall that discussion but was under the impression it revolved around Thread.interrupt() not working on commonPool task submissions.

I was asking if you don't rely on Thread.interrupt(), using WatcherService.close() instead, then is that sufficient for commonPool too.

No worries if it is really insufficient, just making sure. 😄 

For example, the following unit test passes for me.

```
public class WatchServiceTest
{
    @Test
    public void shouldClose() throws Exception
    {
        final WatchService service = FileSystems.getDefault().newWatchService();

        Runnable task = () ->
        {
            watch:
            while (true)
            {
                try
                {
                    service.take();
                }
                catch (InterruptedException | ClosedWatchServiceException ex)
                {
                    break watch;
                }
            }
        };

        final Future<?> future = ForkJoinPool.commonPool().submit(task);

        service.close();

        future.get();
    }
}
```

Reply:
Actually, this looks like it's enoguh. I also tried this method, but what I think I missed is the future.get call. And it really makes sense that without that, it's not working .